# Riiid Data Wrangling

### Table of content<a id="TOC"></a>

* [Data Collection](#Step1)<br>
* [Data Organization](#Step2)<br>
* [Data Definition](#Step3)<br>
* [Data Cleaning](#Step4)<br>
* [Finalize Dataset](#Step5)<br>

## <a class="anchor" id="Step1">[Step one - Data Collection](#TOC)</a>

The first step is to organize the data based on the below three steps:
    1. import all the libraries
    2. Data loading
    3. Data joining

In [1]:
import pandas as pd

In [2]:
path = '/Users/adelsadr/Springboard-Projects/CapstoneTwoRiiid/riiid-test-answer-prediction/'

## <a class="anchor" id="Step2">[Step two - Data Organization](#TOC)</a>

Data organization is about creating a file structure and add all the works to the GitHub repository related to this project.
    1. File structure
    2. GitHub

<font color=blue>This project includes five Jupyter Notebooks each for Wrangling, Exploratory Data Analysis, Preprocessing & Training, and Modeling. 'riiid-test-answer-prediction' folder also includes all the data in csv format and finally there is a deck of slides to explain the story behinde this problem.<br>To keep the record of this project I will use the Git version control. All the codes will push to the repository on GitHub after each changes.<font>

## <a class="anchor" id="Step3">[Step three - Data Definition](#TOC)</a>

THe goal of this step is to gain an understanding of the data features to inform the
next steps of the project.
    1. Column names
    2. Data types
    3. Description of the columns
    4. Counts and percents unique values
    5. Ranges of values

In [3]:
example_sample_submission = pd.read_csv(path + 'example_sample_submission.csv')
example_sample_submission.info()
print('\n\n\n')
print(example_sample_submission.head())
print('\n\n\n')
example_sample_submission.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104 entries, 0 to 103
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   row_id              104 non-null    int64  
 1   answered_correctly  104 non-null    float64
 2   group_num           104 non-null    int64  
dtypes: float64(1), int64(2)
memory usage: 2.6 KB




   row_id  answered_correctly  group_num
0       0                 0.5          0
1       1                 0.5          0
2       2                 0.5          0
3       3                 0.5          0
4       4                 0.5          0






count       mean        std  min    25%   50%    75%  \
row_id              104.0  53.480769  32.012734  0.0  25.75  53.5  80.25   
answered_correctly  104.0   0.500000   0.000000  0.5   0.50   0.5   0.50   
group_num           104.0   1.711538   1.094320  0.0   1.00   2.0   3.00   

                      max  
row_id              108.0  
answered_correctly    0.5  
group_num             3.0

In [4]:
example_sample_submission['answered_correctly'].unique()

array([0.5])

<font color=red><b>QUESTION 1</b><font><br>
<font color=red>answered_correctly is: "(int8) if the user responded correctly. Read -1 as null, for lectures." based on the data description, but here it shows only 0.5 which is nonsense!<font>

In [5]:
example_sample_submission['group_num'].unique()

array([0, 1, 2, 3])

<font color=red><b>QUESTION 2</b><font><br>
<font color=red>I counldn't find anything regarding "group_num" in the data description.<font>

In [6]:
example_sample_submission.duplicated(subset='row_id', keep='first').sum()

0

<font color=blue>There is no duplicate in this dataframe.<font>

In [7]:
example_test = pd.read_csv(path + 'example_test.csv')
example_test.info()
print('\n\n\n')
print(example_test.head())
print('\n\n\n')
example_test.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104 entries, 0 to 103
Data columns (total 11 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   row_id                          104 non-null    int64  
 1   group_num                       104 non-null    int64  
 2   timestamp                       104 non-null    int64  
 3   user_id                         104 non-null    int64  
 4   content_id                      104 non-null    int64  
 5   content_type_id                 104 non-null    int64  
 6   task_container_id               104 non-null    int64  
 7   prior_question_elapsed_time     103 non-null    float64
 8   prior_question_had_explanation  103 non-null    object 
 9   prior_group_answers_correct     4 non-null      object 
 10  prior_group_responses           4 non-null      object 
dtypes: float64(1), int64(7), object(3)
memory usage: 9.1+ KB




   row_id  group_num    timestamp   

count          mean           std        min  \
row_id                       104.0  5.348077e+01  3.201273e+01        0.0   
group_num                    104.0  1.711538e+00  1.094320e+00        0.0   
timestamp                    104.0  1.669283e+10  2.039655e+10        0.0   
user_id                      104.0  1.156358e+09  5.966096e+08  7792299.0   
content_id                   104.0  6.772702e+03  4.503711e+03       18.0   
content_type_id              104.0  0.000000e+00  0.000000e+00        0.0   
task_container_id            104.0  1.094413e+03  1.588054e+03        0.0   
prior_question_elapsed_time  103.0  2.494963e+04  9.969309e+03     7000.0   

                                      25%           50%           75%  \
row_id                       2.575000e+01  5.350000e+01  8.025000e+01   
group_num                    1.000000e+00  2.000000e+00  3.000000e+00   
timestamp                    1.574161e+09  4.693169e+09  2.746557e+10   
user_id                      6.448233e+08  1.257605e+09  1.609175e+09   
content_id                   2.028750e+03  6.225500e+03  1.103125e+04   
content_type_id              0.000000e+00  0.000000e+00  0.000000e+00   
task_container_id            1.627500e+02  3.785000e+02  1.179000e+03   
prior_question_elapsed_time  1.800000e+04  2.466700e+04  2.887500e+04   

                                      max  
row_id                       1.080000e+02  
group_num                    3.000000e+00  
timestamp                    7.681646e+10  
user_id                      2.103437e+09  
content_id                   1.335400e+04  
content_type_id              0.000000e+00  
task_container_id            6.958000e+03  
prior_question_elapsed_time  7.240000e+04

In [8]:
example_test.duplicated(subset='row_id', keep='first').sum()

0

<font color=blue>There is no duplication in this dataframe.<font>

In [9]:
example_test.loc[example_test['prior_group_answers_correct'].notnull()]

row_id  group_num  timestamp    user_id  content_id  content_type_id  \
0        0          0          0  275030867        5729                0   
18      18          1      23609  275030867        5502                0   
45      46          2      42512  275030867        6133                0   
71      74          3      75311  275030867        8308                0   

    task_container_id  prior_question_elapsed_time  \
0                   0                          NaN   
18                  1                      34000.0   
45                  2                      17000.0   
71                  3                      15000.0   

   prior_question_had_explanation  \
0                             NaN   
18                          False   
45                          False   
71                          False   

                          prior_group_answers_correct  \
0                                                  []   
18  [0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, ...   
45  [1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, ...   
71  [1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, ...   

                                prior_group_responses  
0                                                  []  
18  [0, 0, 1, 1, 0, 1, 3, 3, 1, 1, 0, 3, 1, 2, 2, ...  
45  [1, 2, 2, 2, 1, 0, 1, 2, 1, 1, 3, 2, 0, 0, 3, ...  
71  [0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 3, 0, 0, ...

In [10]:
set((example_test['prior_group_answers_correct'].loc[example_test['row_id']==18]).iloc[0])

{' ', ',', '0', '1', '[', ']'}

In [11]:
set((example_test['prior_group_answers_correct'].loc[example_test['row_id']==46]).iloc[0])

{' ', ',', '0', '1', '[', ']'}

In [12]:
set((example_test['prior_group_answers_correct'].loc[example_test['row_id']==74]).iloc[0])

{' ', ',', '0', '1', '[', ']'}

In [13]:
set((example_test['prior_group_responses'].loc[example_test['row_id']==18]).iloc[0])

{' ', ',', '0', '1', '2', '3', '[', ']'}

In [14]:
set((example_test['prior_group_responses'].loc[example_test['row_id']==46]).iloc[0])

{' ', ',', '0', '1', '2', '3', '[', ']'}

In [15]:
set((example_test['prior_group_responses'].loc[example_test['row_id']==74]).iloc[0])

{' ', ',', '0', '1', '3', '[', ']'}

<font color=red><b>QUESTION 3</b><font><br>
<font color=red>I don't understand these two columns! why all rows are null except four of them? should I keep them? how can I use them?<font>

In [16]:
example_test['prior_question_had_explanation'].unique()

array([nan, True, False], dtype=object)

In [17]:
example_test['prior_question_had_explanation'].isnull().sum()

1

In [18]:
example_test['user_id'].nunique()

42

<font color=blue>This dataframe represents 42 users and one of the rows with the null value is the user's first question bundle.<font>

In [19]:
example_test['content_id'].nunique()

103

In [20]:
example_test['content_type_id'].value_counts()

0    104
Name: content_type_id, dtype: int64

In [21]:
example_test['task_container_id'].nunique()

74

<font color=blue>104 user interaction with 103 questions being posed to users in 74 batches, and not any lectures.<font>

In [22]:
lectures = pd.read_csv(path + 'lectures.csv')
lectures.info()
print('\n\n\n')
print(lectures.head())
print('\n\n\n')
lectures.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   lecture_id  418 non-null    int64 
 1   tag         418 non-null    int64 
 2   part        418 non-null    int64 
 3   type_of     418 non-null    object
dtypes: int64(3), object(1)
memory usage: 13.2+ KB




   lecture_id  tag  part           type_of
0          89  159     5           concept
1         100   70     1           concept
2         185   45     6           concept
3         192   79     5  solving question
4         317  156     5  solving question






count          mean          std   min      25%      50%  \
lecture_id  418.0  16983.401914  9426.164660  89.0  9026.25  17161.5   
tag         418.0     94.480861    53.586487   0.0    50.25     94.5   
part        418.0      4.267943     1.872424   1.0     2.00      5.0   

                 75%      max  
lecture_id  24906.25  32736.0  
tag           140.00    187.0  
part            6.00      7.0

In [23]:
lectures['type_of'].unique()

array(['concept', 'solving question', 'intention', 'starter'],
      dtype=object)

In [24]:
lectures['tag'].nunique()

151

In [25]:
lectures['lecture_id'].nunique()

418

In [26]:
lectures['part'].value_counts()

5    143
6     83
2     56
1     54
7     32
4     31
3     19
Name: part, dtype: int64

<font color=blue>This dataframe includes 418 lectures with four different core purpose, in 151 clusters. Lectures are mostly concentrated on category 5.<font>

In [27]:
questions = pd.read_csv(path + 'questions.csv')
questions.info()
print('\n\n\n')
print(questions.head())
print('\n\n\n')
questions.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13523 entries, 0 to 13522
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   question_id     13523 non-null  int64 
 1   bundle_id       13523 non-null  int64 
 2   correct_answer  13523 non-null  int64 
 3   part            13523 non-null  int64 
 4   tags            13522 non-null  object
dtypes: int64(4), object(1)
memory usage: 528.4+ KB




   question_id  bundle_id  correct_answer  part            tags
0            0          0               0     1   51 131 162 38
1            1          1               1     1       131 36 81
2            2          2               0     1  131 101 162 92
3            3          3               0     1  131 149 162 29
4            4          4               3     1    131 5 162 38






count         mean          std  min     25%     50%  \
question_id     13523.0  6761.000000  3903.898180  0.0  3380.5  6761.0   
bundle_id       13523.0  6760.510907  3903.857783  0.0  3379.5  6761.0   
correct_answer  13523.0     1.455298     1.149707  0.0     0.0     1.0   
part            13523.0     4.264956     1.652553  1.0     3.0     5.0   

                    75%      max  
question_id     10141.5  13522.0  
bundle_id       10140.0  13522.0  
correct_answer      3.0      3.0  
part                5.0      7.0

<font color=blue>There are 13523 questions in 9765 batches. correct answers are provided without any null value to compare to 'user_answer' from the train dataset.<br>Most of the questions are concentrated on section 5 of the TOEIC test.<br>'tags' column has one null record, and unlike the lectures dataset, questions can have one or more tag code. There are 1519 clusters of question in questions dataset.<font>

In [28]:
questions['question_id'].nunique()

13523

In [29]:
questions['bundle_id'].nunique()

9765

In [30]:
questions['correct_answer'].value_counts()

0    3716
3    3544
1    3478
2    2785
Name: correct_answer, dtype: int64

In [31]:
questions['part'].value_counts()

5    5511
2    1647
3    1562
4    1439
6    1212
7    1160
1     992
Name: part, dtype: int64

In [32]:
questions['tags'].nunique()

1519

In [33]:
questions[questions['tags'].isnull()]

question_id  bundle_id  correct_answer  part tags
10033        10033      10033               2     6  NaN

In [34]:
train = pd.read_csv(path + 'train.csv')
train.info()
print('\n\n\n')
print(train.head())
print('\n\n\n')
train.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101230332 entries, 0 to 101230331
Data columns (total 10 columns):
 #   Column                          Dtype  
---  ------                          -----  
 0   row_id                          int64  
 1   timestamp                       int64  
 2   user_id                         int64  
 3   content_id                      int64  
 4   content_type_id                 int64  
 5   task_container_id               int64  
 6   user_answer                     int64  
 7   answered_correctly              int64  
 8   prior_question_elapsed_time     float64
 9   prior_question_had_explanation  object 
dtypes: float64(1), int64(8), object(1)
memory usage: 7.5+ GB




   row_id  timestamp  user_id  content_id  content_type_id  task_container_id  \
0       0          0      115        5692                0                  1   
1       1      56943      115        5716                0                  2   
2       2     118363      115     

count          mean           std    min  \
row_id                       101230332.0  5.061517e+07  2.922268e+07    0.0   
timestamp                    101230332.0  7.703644e+09  1.159266e+10    0.0   
user_id                      101230332.0  1.076732e+09  6.197163e+08  115.0   
content_id                   101230332.0  5.219605e+03  3.866359e+03    0.0   
content_type_id              101230332.0  1.935222e-02  1.377596e-01    0.0   
task_container_id            101230332.0  9.040624e+02  1.358302e+03    0.0   
user_answer                  101230332.0  1.376123e+00  1.192896e+00   -1.0   
answered_correctly           101230332.0  6.251644e-01  5.225307e-01   -1.0   
prior_question_elapsed_time   98878794.0  2.542381e+04  1.994815e+04    0.0   

                                      25%           50%           75%  \
row_id                       2.530758e+07  5.061517e+07  7.592275e+07   
timestamp                    5.243436e+08  2.674234e+09  9.924551e+09   
user_id                      5.408116e+08  1.071781e+09  1.615742e+09   
content_id                   2.063000e+03  5.026000e+03  7.425000e+03   
content_type_id              0.000000e+00  0.000000e+00  0.000000e+00   
task_container_id            1.040000e+02  3.820000e+02  1.094000e+03   
user_answer                  0.000000e+00  1.000000e+00  3.000000e+00   
answered_correctly           0.000000e+00  1.000000e+00  1.000000e+00   
prior_question_elapsed_time  1.600000e+04  2.100000e+04  2.966600e+04   

                                      max  
row_id                       1.012303e+08  
timestamp                    8.742577e+10  
user_id                      2.147483e+09  
content_id                   3.273600e+04  
content_type_id              1.000000e+00  
task_container_id            9.999000e+03  
user_answer                  3.000000e+00  
answered_correctly           1.000000e+00  
prior_question_elapsed_time  3.000000e+05

In [35]:
train.isnull().sum()

row_id                                  0
timestamp                               0
user_id                                 0
content_id                              0
content_type_id                         0
task_container_id                       0
user_answer                             0
answered_correctly                      0
prior_question_elapsed_time       2351538
prior_question_had_explanation     392506
dtype: int64

<font color=blue>Null values in these two columns show the user's first question bundle or lecture, so it is ok to keep these records.<font>

## <a class="anchor" id="Step4">[Step four - Data Cleaning](#TOC)</a>

Clean up the data in order to prepare it for the next steps of
your project.
    1. NA or missing values
    2. Duplicates

<font color=blue>All the csv files has been checked, these datasets are clean enough to use.<font>

## <a class="anchor" id="Step5">[Step five - Finalize Dataset](#TOC)</a>

Save the clean data to a csv file to use in the future processes.

<font color=blue>Final csv files are saved and available in 'riiid-test-answer-prediction' folder<font>